docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -73,27 +74,57 @@ and includes all the core field members that are common across all transaction t
 Developers do not have to include `StdTxBase` if they wish, so it is meant to be
 used as an auxiliary type.
 
-### Signing
+### Signing & Transaction Submission
 
-Signing of a `Transaction` must be canonical across clients and binaries. In order
-to provide canonical representation of a `Transaction` to sign over, clients must
+To provide an acceptable user experience for client-side developers, we separate
+transaction encoding and signing. While every app defines its own transaction
+types for encoding purposes, we define a single standard signing `SignDoc` to be used
+across apps that leverages `google.protobuf.Any` to handle interface types. `Any`
+can be used to wrap any protobuf message by encoding both its type URL and value.
+We avoid using `Any` during the encoding phase because these URLs can be long
+and we want to keep encoded messages small. We are not, however, sending the
+`SignDoc` (encoded in JSON) over the wire or storing it on nodes,
+therefore using `Any` for signing does not add any encoding overhead.
+
+```proto
+message SignDoc {
+  StdSignDocBase base = 1;
+  repeated google.protobuf.Any msgs = 2;
+}
+```
+
+Signing of a `SignDoc` must be canonical across clients and binaries. In order
+to provide canonical representation of a `SignDoc` to sign over, clients must
 obey the following rules:
 
-- Encode `SignDoc` (see below) via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
+- Encode `SignDoc` via [Protobuf's canonical JSON encoding](https://developers.google.com/protocol-buffers/docs/proto3#json).
   - Default must be stripped from the output!
   - JSON keys adhere to their Proto-defined field names.
 - Generate canonical JSON to sign via the [JSON Canonical Form Spec](https://gibson042.github.io/canonicaljson-spec/).
   - This spec should be trivial to interpret and implement in any language.
 
-```Protobuf
-// app/codec/codec.proto
+Because signing and encoding are separated and apps will know how to encode
+a signing messages, we can provide a gRPC/REST endpoint that allows for generic
+app-independent transaction submission. This can be done via a generic
+`AnyTransaction` type:
 
-message SignDoc {
-  StdSignDocBase base = 1;
-  repeated Message msgs = 2;
+```proto
+message AnyTransaction {
+  cosmos_sdk.codec.std.v1.StdTxBase base = 1;
+  repeated google.protobuf.Any msgs = 2;
 }
 ```
 
+On the backend, this will be converted to the app-specific `Transaction` type,
+with any errors returned for unsupported `Msg`s. Client developers that prefer
+to use the app-specific `Transaction` and `Message` `oneof`s can also do this
+encoding on their own. This provides additional type safety where needed. There
+is, however, no security disadvantage for using the generic RPC transaction
+endpoint because all clients regardless of whether they have access to the
+app-level proto files will know how to correctly sign transactions. This allows
```

Comment:
Without app-level proto files, it is not possible to get the JSON representation from a message of type `google.protobuf.Any`. For this reason, a generic client can neither sign nor verify transactions without app-level proto files.

Reply:
That's not true. They just need the module level proto files. That's what this PR accomplishes. It makes having the app level proto files optional for clients

Reply:
Oh, interesting, thanks. Makes sense.

I was not aware that there are module-level proto files and app-level proto files. But yeah, it is a consequence of how the protobuf messages are defined in this document. Since this differentiation seems to be a core concept of this PR, I think it would be good to mention it explicitely. This also answers the logistics question from above: there is a limited set of standard module protobuf files that is sufficient to interact with any chain, as long as no module is changed or custom modules are added.

Reply:
I'd recommend going through both Proto ADRs @webmaster128 and taking a look at the current `master` of the SDK. You'll see how the module-level proto types are defined and the app-level proto types are defined. In addition, you'll see how the codec is defined and used to be modular.

Reply:
> This also answers the logistics question from above: there is a limited set of standard module protobuf files that is sufficient to interact with any chain, as long as no module is changed or custom modules are added.

Yes, this PR effectively allows a client to broadcast transactions to a chain that implements standard modules without needing to have its app-level proto file. So if I write a client app for the `gov` module say, I can support new chains just by knowing their chain-id and having access to a node. Without this PR, I would need to do code generation on the app-level proto file plus write some adapter code for every new chain that my client app supports. That's the UX difference. Is the context this now clearer?

I agree that reviewing the current approach on master would help give context and since you are a client developer representative here @webmaster128 we highly value your input regarding the different approaches under consideration.